altair/utils/core.py

```diff
@@ -767,7 +769,132 @@ def display_traceback(in_ipython: bool = True):
         traceback.print_exception(*exc_info)
 
 
-def infer_encoding_types(args: Sequence, kwargs: MutableMapping, channels: ModuleType):
+_ChannelType = Literal["field", "datum", "value"]
+_CHANNEL_CACHE: "_ChannelCache"
+"""Singleton `_ChannelCache` instance.
+
+Initialized on first use.
+"""
+
+
+class _ChannelCache:
+    channel_to_name: Dict[Type[SchemaBase], str]
+    name_to_channel: Dict[str, Dict[_ChannelType, Type[SchemaBase]]]
+
+    @classmethod
+    def from_channels(cls, channels: ModuleType, /) -> "_ChannelCache":
+        # - This branch is only kept for tests that depend on mocking `channels`.
+        # - No longer needs to pass around `channels` reference and rebuild every call.
+        c_to_n = {
+            c: c._encoding_name
+            for c in channels.__dict__.values()
+            if isinstance(c, type)
+            and issubclass(c, SchemaBase)
+            and hasattr(c, "_encoding_name")
+        }
+        self = cls.__new__(cls)
+        self.channel_to_name = c_to_n
+        self.name_to_channel = _invert_group_channels(c_to_n)
+        return self
+
+    @classmethod
+    def from_cache(cls) -> "_ChannelCache":
+        global _CHANNEL_CACHE
+        try:
+            cached = _CHANNEL_CACHE
+        except NameError:
+            cached = cls.__new__(cls)
+            cached.channel_to_name = _init_channel_to_name()
+            cached.name_to_channel = _invert_group_channels(cached.channel_to_name)
+            _CHANNEL_CACHE = cached
+        return _CHANNEL_CACHE
+
+    def get_encoding(self, tp: Type[Any], /) -> str:
+        if encoding := self.channel_to_name.get(tp):
+            return encoding
+        raise NotImplementedError(f"positional of type {type(tp).__name__!r}")
+
+    def _wrap_in_channel(self, obj: Any, encoding: str, /):
```

Comment:
It is the first time I see a forward slash, `/` as argument within a function. Can you explain what that does?

Reply:
Sure @mattijn 

`/` is a marker for [positional-only parameters](https://peps.python.org/pep-0570/#benefits-of-positional-only-parameters). Any parameter to the left of `/` is positional-only, and may not be used by name.
In this case, calling `self._wrap_in_channel(encoding="enc", obj=[1,2,3])` raises a `TypeError`. The [python tutorial docs](https://docs.python.org/3/tutorial/controlflow.html#special-parameters) may be helpful for an overview as well. 

Personally, I like to use `/` in cases like:
- The function/method is not part of the public API, to leave more flexibility in renaming parameters in the future, without introducing a breaking change
- There are a 1-3 parameters
- The function is *currently* used in one specific way and the function name and parameter order are clear
	- In this case, `_wrap_in_channel` could logically be thought of as having parameters `_wrap_in_channel(wrappee, wrapper)`.

Looking back at PEP570, I see that this was introduced in `python3.8`, which may explain the feature's absence in `altair` until now.

Hope all of that was helpful